### ISSUE TITLE
stacks does not recognize control argument when adding candidate

### ISSUE THREAD

**Issue by amin0511ss** [tidymodels/stacks#194]
I am trying to stack two models: ranger (randomeforest) and lighgbm from bonsai. I fit resamples via respective workflows and include the control_stack_resamples argument. Both workflows use the same recipe as follows.
receipe_mixed_undersample <-
recipe(Disposition ~ ., data = Data202101_train) %>%
step_lencode_mixed(RuleCombination, outcome = vars(Disposition)) %>%
step_nzv( all_predictors(), options = list(freq_cut = 95/5, unique_cut = 10 )) %>%
step_corr( all_numeric(), - all_outcomes(), threshold = 0.9 ) %>%
step_downsample (Disposition, under_ratio = 1) %>%
step_cut(BankCustomer, breaks = 1)

cv folds:
resample_folds <- vfold_cv(Data202101_train, v=5,repeats = 3, strata = Disposition)

workflows and fit:
stacks_lgbm_mixed_undersample_workflow <-
workflow() %>% add_recipe(receipe_mixed_undersample) %>% add_model(last_lgbm_mixed_undersample_model)

lgbm_res <-
stacks_lgbm_mixed_undersample_workflow %>%
fit_resamples(resamples = resample_folds,metrics = metric_set(mn_log_loss,mcc,bal_accuracy,roc_auc),
control_stack_resamples(save_pred = TRUE, save_workflow = TRUE))

stacks_rf_mixed_undersample_workflow <-
workflow() %>% add_recipe(receipe_mixed_undersample) %>% add_model(last_rf_mixed_undersample_model)

rf_res <-
stacks_rf_mixed_undersample_workflow %>%
fit_resamples(resamples = resample_folds,metrics = metric_set(mn_log_loss,mcc,bal_accuracy,roc_auc),
control_stack_resamples(save_pred = TRUE, save_workflow = TRUE))

The models run fine and produce result but when i try to add candidate, i get the error message below:

My_data_st <-
stacks() %>%
add_candidates(lgbm_res) %>%
add_candidates(rf_res)

Error in add_candidates():
! The inputted candidates argument was not generated with the appropriate control settings. Please
see control_stack().
Run rlang::last_error() to see where the error occurred.

Could someone please advise as to what i need to do.
thank you

**Comment by simonpcouch** [comment]
Thanks for the issue!

Can you please provide a minimal [reprex](https://github.com/tidyverse/reprex#what-is-a-reprex) (reproducible example)? I don't have access to your `Data202101_train` data, so am unable to reproduce your result. A reprex will help me troubleshoot and fix your issue more quickly.🙂

**Comment by amin0511ss** [comment]
Thank you. I will try to do it by EOD. In the meantime, is there anything in the code that you would suspect is causing the issue. The dataset:
1. Binary classification outcome (0, 1)
2. Before recipe: 146 features, 2 categorical and the rest numeric. One categorical has two levels, the other one has more than 400 levels. this was is target encoded in the recipe
3. after recipe: 65 features, 1 categorical and the rest numeric.

thank you

**Comment by amin0511ss** [comment]
i created a reproducible code which included a dataset with 186 rows and 146 columns. but it does not allow me to submit it as it exceeds maximum number of characters.

**Comment by amin0511ss** [comment]
i cann't make the dataset small enough to paste it here. Even with 185 rows of data, it would exceed maximum allowable limit. Here is the code beyond the dataset. If make the dataset smaller then it get the singularity error message when i run the models. Please take a look at the code and let me know if you can see something that would cause the algorithm not recognize the control statement in the add candidate step.


recipe_0 <- 
  recipe(X_1 ~ ., data = DF) %>%
  step_lencode_mixed(X_145, outcome = vars(X_1)) %>% 
  step_nzv( all_predictors(), options = list(freq_cut = 95/5, unique_cut = 10 )) %>%
  step_corr( all_numeric(), - all_outcomes(), threshold = 0.9 ) %>%
  step_downsample (X_1, under_ratio = 1) 

recipe_1 <- recipe_0 %>%
  step_cut(X_130, breaks = 1) 

library(bonsai)
library(yardstick)
library(dials)

lighgbm_model <- boost_tree(
  trees = 1000,  # Trees (type: integer, default: 100)
  tree_depth = tune(), # Tree Depth (type: integer, default: -1)
  learn_rate = tune(), # : Learning Rate (type: double, default: 0.1)
  mtry = tune(), # : # Randomly Selected Predictors (type: integer, default: see below)
  min_n = tune(), # : Minimal Node Size (type: integer, default: 20)
  loss_reduction = tune(), # : Minimum Loss Reduction (type: double, default: 0)
  stop_iter = 5
) %>%
  set_engine(engine ='lightgbm') %>%
  set_mode(mode = "classification") 


lgbm_param <- extract_parameter_set_dials(lighgbm_model) %>% update(mtry = mtry(c(6,18)))
lgbm_param

lgbm_grid <- dials:::grid_max_entropy.parameters(lgbm_param, size = 40)

########### randomForest
library(parsnip)
rf_model <- rand_forest(
  mtry = tune(),
  trees = 1001,
  min_n = tune()
) %>%
  set_engine("ranger") %>%
  set_mode("classification") 

rf_param <- extract_parameter_set_dials(rf_model) %>% update(mtry = mtry(c(6,18)))
rf_param

rf_grid <- dials:::grid_max_entropy.parameters(rf_param, size = 40)

########### xgboost
xgb_model <- boost_tree(
  trees = 1000,
  tree_depth = tune(), min_n = tune(),
  loss_reduction = tune(),                     ## first three: model complexity
  sample_size = tune(), mtry = tune(),         ## randomness
  learn_rate = tune()                          ## step size
) %>%
  set_engine("xgboost") %>%
  set_mode("classification")

xgb_param <- extract_parameter_set_dials(xgb_model) %>% update(mtry = mtry(c(6,18)))


xgb_grid <- dials:::grid_max_entropy(xgb_param, size = 40)


xgb_workflow <- workflow() %>%  add_recipe(recipe_0) %>%
  add_model(xgb_model)

lgbm_workflow <- workflow() %>%  add_recipe(recipe_1) %>%
  add_model(lighgbm_model)

rf_workflow <- workflow() %>%  add_recipe(recipe_1) %>%
  add_model(rf_model)


DF_cv_folds <-  vfold_cv(DF, v=2,repeats = 2, strata = X_1)


xgb_stacks_grid_tuned <-  tune_grid(
  xgb_workflow,
  resamples = DF_cv_folds,
  xgb_grid1,
  metrics = metric_set(mn_log_loss,mcc,bal_accuracy,roc_auc),
  control_stack_grid(save_pred = TRUE)
)

rf_stacks_grid_tuned <-  tune_grid(
  lgbm_workflow,
  resamples = DF_cv_folds,
  rf_grid,
  metrics = metric_set(mn_log_loss,mcc,bal_accuracy,roc_auc),
  control_stack_grid(save_pred = TRUE)
)


lgbm_stacks_grid_tuned <-  tune_grid(
  lgbm_workflow,
  resamples = DF_cv_folds,
  lgbm_grid,
  metrics = metric_set(mn_log_loss,mcc,bal_accuracy,roc_auc),
  control_grid(save_pred = TRUE)
)


ISP_data_st <- 
  stacks() %>%
  add_candidates(xgb_stacks_grid_tuned) %>%
  add_candidates(rf_stacks_grid_tuned) %>%
  add_candidates(lgbm_stacks_grid_tuned)

**Comment by amin0511ss** [comment]
I removed the argument but still get the same annoying message. the tune grid steps run without any error messages but the stacks does not work

  xgb_stacks_grid_tuned <-  tune_grid(
    xgb_mixed_workflow,
    resamples = cv_folds,
    xgb_grid1,
    metrics = metric_set(mn_log_loss,mcc,bal_accuracy,roc_auc),
    control_stack_grid()
  )
  
  rf_stacks_grid_tuned <-  tune_grid(
    rf_mixed_undersample_workflow,
    resamples = cv_folds,
    rf_grid,
    metrics = metric_set(mn_log_loss,mcc,bal_accuracy,roc_auc),
    control_stack_grid()
  )
  
  lgbm_stacks_grid_tuned <-  tune_grid(
    lgbm_mixed_undersample_workflow,
    resamples = cv_folds,
    lgbm_grid,
    metrics = metric_set(mn_log_loss,mcc,bal_accuracy,roc_auc),
    control_stack_grid()
  )
  
  ISP_data_st <- 
    stacks() %>%
    add_candidates(xgb_stacks_grid_tuned) %>%
    add_candidates(rf_stacks_grid_tuned) %>%
    add_candidates(lgbm_stacks_grid_tuned)

Error in `add_candidates()`:
! The inputted `candidates` argument was not generated with the appropriate control settings. Please see
  [`control_stack()`](stacks::control_stack).
